### ISSUE TITLE
`nan` gradient when `tf.where` is used

### ISSUE THREAD

**Issue by 0x0badc0de** [tensorflow/tensorflow#38349]
<em>Please make sure that this is a bug. As per our
[GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md),
we only address code/doc bugs, performance issues, feature requests and
build/installation issues on GitHub. tag:bug_template</em>

**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow):  Yes
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): Debian GNU/Linux 10 (buster)
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if
the issue happens on mobile device: 
- TensorFlow installed from (source or
binary): binary
- TensorFlow version (use command below): v2.1.0-rc2-17-ge5bf8de 2.1.0 / v1.12.1-29016-g38797a1c8b 2.2.0-dev20200407
- Python version: 3.7.7
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source): 
- CUDA/cuDNN version: - GPU model and memory:

You can collect some of this information using our environment capture
[script](https://github.com/tensorflow/tensorflow/tree/master/tools/tf_env_collect.sh)
You can also obtain the TensorFlow version with: 1. TF 1.0: `python -c "import
tensorflow as tf; print(tf.GIT_VERSION, tf.VERSION)"` 2. TF 2.0: `python -c
"import tensorflow as tf; print(tf.version.GIT_VERSION, tf.version.VERSION)"`

**Describe the current behavior**
Well-defined function with `tf.where` has `nan` gradients at points where `tf.where` inactive branch is undefined.

**Describe the expected behavior**
Inactive branch should be ignored in gradients calculations.

**Standalone code to reproduce the issue** 
Provide a reproducible test case that is the bare minimum necessary to generate
the problem. If possible, please share a link to Colab/Jupyter/any notebook.
```
import tensorflow as tf

for ex in range(-3, 3):
    x = tf.convert_to_tensor(10.**ex)
    with tf.GradientTape() as g:
        g.watch(x)
        y = tf.where(x >= -1., x, tf.math.log1p(-x))
#         y = tf.where(x >= -1., x, tf.math.log(1.-x))
#         y = tf.where(x >= -1., x, 1./(1.-x))
    dy_dx = g.gradient(y, x)
    print(f'y({x})={y}, dy/dx({x})={dy_dx}')
```

All 3 functions above are well defined for positive values used for testing. Still they show no gradient at point `1.`. while it has to be equal to `1.`

```
y(0.0010000000474974513)=0.0010000000474974513, dy/dx(0.0010000000474974513)=1.0
y(0.009999999776482582)=0.009999999776482582, dy/dx(0.009999999776482582)=1.0
y(0.10000000149011612)=0.10000000149011612, dy/dx(0.10000000149011612)=1.0
y(1.0)=1.0, dy/dx(1.0)=nan
y(10.0)=10.0, dy/dx(10.0)=1.0
y(100.0)=100.0, dy/dx(100.0)=1.0
```

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

**Comment by ravikyram** [comment]
I have tried on colab with TF version 2.1.0 , 2.2.0-rc2 and was able to reproduce the issue.Please, find the gist [here](https://colab.research.google.com/gist/ravikyram/806f63f2cf04070a4601289d7003cf0a/untitled24.ipynb). Thanks!

**Comment by mdanatg** [comment]
This is due to a limitation limitation in how gradients are calculated. Unfortunately, it is unlikely to be fixed in the foreseable future.

You can find more detail here, along with a recipe for how to avoid it: https://stackoverflow.com/questions/33712178/tensorflow-nan-bug/42497444#42497444

In short, if the input to a tf.where contains NaNs, the gradient will always be NaN, regardless whether the input is actually used or not, and the workaround is to prevent the inputs from ever containing NaNs.

**Comment by 0x0badc0de** [comment]
Shouldn't this be documented with big warning in `tf.where` docs in this case?

**Comment by mdanatg** [comment]
Indeed it should.

**Comment by joemaren** [comment]
@mdanatg Hello, this is my first time contributing to TensofFlow lib. From the thread I gather you would require the `tf.where` be updated. If it is so can I work on this?

**Comment by anorak-k** [comment]
Should the updated doc contain a something like a warning? or will a small note at the end about the input not being Nan will do? Also should the workaround for avoiding it also be added to the doc?

**Comment by mdanatg** [comment]
@joemaren @anorak-k 

Sorry for the delay. Feel free to send a PR - it's only a matter of adding a paragraph to the docstring.

It text should be more in the lines of a warning. Something like: **Important: if any of the inputs have NaN values, etc.**. And yes, it should include the workaround as well, which is something in the lines of: instead of `tf.where(x, ops_that_can_nan(z), ...)`, write `tf.where(x, ops_that_can_nan(tf.where(x, z, safe_value)), ...)`.

**Comment by anorak-k** [comment]
@mdanatg I have added the change and raised a PR #38467

**Comment by mkaze** [comment]
@mdanatg Thanks for your reply. However, I would like to mention that this behavior also happens when the generated value in the inactive branch **is not finite** (i.e. `inf` or `-inf`). Here is a minimal reproducible example:

```python
import tensorflow as tf

a = tf.Variable(10.)
with tf.GradientTape() as tape:
  out = tf.where(a < 15., a, tf.math.pow(10.0, tf.math.exp(a)))
  grads = tape.gradient(out, a)

print(grads)
# tf.Tensor(nan, shape=(), dtype=float32)
```

And also if we reverse the condition such that the branch with infinite value is selected, the gradient would be infinite (which is a bit surprising that it does not generate `nan` instead, like above):
```python
with tf.GradientTape() as tape:
  out = tf.where(a > 15., a, tf.math.pow(10.0, tf.math.exp(a)))
  grads = tape.gradient(out, a)

print(grads)
# tf.Tensor(inf, shape=(), dtype=float32)
```
So this behavior happens for both `nan` and **infinite values** in inactive branch. I wish it wasn't like this, because it's a bit unreasonable (unfortunately, even gradient clipping does not help with this, because clipping a `nan` value produces `nan` in TF).

CC: @anorak-k for potential consideration in your PR after @mdanatg confirms this.

**Comment by mdanatg** [comment]
@mkaze that's true - nan, inf and any other special FP value will disrupt the gradient calculation.

What happens internally is that the gradients are aggregated in this fashion: `1 * <grad of branch taken> + 0 * <grad of branch not taken>`. In the former case, you have `0 * inf = nan`. In the latter case, you have `1 * inf = inf`. I agree it's very confusing, unfortunately a naive fix would add significant overhead to gradient calculations.

Moreover, the forward calculation doesn't need to result in a nan or inf. You can also get weird results if the gradient alone is nan or inf. For example, the cube root function is defined and well-behaved everywhere, but its derivative at zero is infinite. So this will give you a nan gradient too:

```
a = tf.Variable(0.0)
with tf.GradientTape() as tape:
  out = tf.where(a < 1, a, tf.pow(a, 1.0/3.0))
  grads = tape.gradient(out, a)
print(grads)
```

I think the tf.where workaround is useful with infinite values as well, so long as the branch not taken is forced to take a gradient that can be safely multiplied by 0. For your example, it would be something like this:

```
dummy_safe_value = 0
safe_a = tf.where(a > 15., dummy_safe_value, a)
out = tf.where(a > 15., a, tf.math.pow(10.0, tf.math.exp(safe_a)))
```

I agree that it sometimes can be impractical to do, but in principle it should always be possible as long as you control the inputs to the sensitive functions - all they have to do is force finite values in all the elements that are dropped.

**Comment by kari-554** [comment]
I want to fix the issue [#38349](https://github.com/tensorflow/tensorflow/issues/38349)

**Comment by tushar-dalal** [comment]
> This is due to a limitation limitation in how gradients are calculated. Unfortunately, it is unlikely to be fixed in the foreseable future.
> 
> You can find more detail here, along with a recipe for how to avoid it: https://stackoverflow.com/questions/33712178/tensorflow-nan-bug/42497444#42497444
> 
> In short, if the input to a tf.where contains NaNs, the gradient will always be NaN, regardless whether the input is actually used or not, and the workaround is to prevent the inputs from ever containing NaNs.

You can simply have it raise a value error if its getting Nan inputs. Or does it not work like that?

**Comment by unicorn-io** [comment]
Can I work on this issue if someone isn't now?

**Comment by mdanatg** [comment]
@tushar-dalal The challenge is that verifying for such NaN inputs can be taking on performance. When debugging, `tf.debugging.check_numerics` can indeed help with that.

@unicorn-io Feel free to tackle it, but note that it's extremely challenging to solve. That said, there was a PR (#38467) to add a warning message to the docs of tf.where, it would be useful to revive it.

**Comment by unicorn-io** [comment]
I am motivated to do this can you give me some tips to start with I will try my best to understand and resolve this issue.

**Comment by unicorn-io** [comment]
> I am motivated to do this can you give me some tips to start with I will try my best to understand and resolve this issue. @mdanatg

**Comment by mdanatg** [comment]
@unicorn-io You can start by looking at the [gradient code](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/c/eager/tape.h#L149) and understandign how it works. Then you can reproduce when happens in the case of a where with bad gradients.

**Comment by unicorn-io** [comment]
Cool I'll get to it

**Comment by madamalarevanth** [comment]
Hey i would like to work on it. can also help please

**Comment by AbhinavTalari** [comment]
> 
> 
> Cool I'll get to it

This bug is unfortunate unfixale as of now it seems.

**Comment by mdanatg** [comment]
It's indeed very challenging to fix. However, the documentation of affected ops, like `tf.where` can still be updated to alert the users about it.

**Comment by unicorn-io** [comment]
@mdanatg isn't #38497 addressing this and is closed?

**Comment by mdanatg** [comment]
You mean #38467? It's closed due to staleness, and it would be useful to revive. By the looks of it it's safe to assume noone else is working on it.

**Comment by EbiereVO** [comment]
Seems like its a long time since the last activity. Is this issue still open to be worked on?

**Comment by mdanatg** [comment]
I think so. There are two parts to it: (1) updating the docs of tf.where, which is fairly straightforward, and (2) actually trying to address the issue, which is a significant undertaking because it involves a rather fundamental issue.

**Comment by iamharshit13** [comment]
Is this issue still addressable ?

**Comment by codeadmin-peritiae** [comment]
Nice to be part of the group.
Please, have a look to my pull request for the workaround: #41721 
I'm going to work on the main issue too.

**Comment by mdanatg** [comment]
@codeadmin-peritiae The PR appears to be empty. Perhaps there's an issue with the git client?

**Comment by Harsh188** [comment]
Just to follow up on the events. It looks like codeadmin-peritiae had an issue with his original PR #41721 where he had trouble with his SSH certificate. He then opened up another PR #41775 which is currently blocked since some of the checks haven't completed. By the looks of it, the documentation update part of this problem is almost completed.

**Comment by mandar1110** [comment]
I am looking for Risk Product manager & Hyderabad with 10-13 yrs experience. For this profile I am looking for candidates who have started their career in coding/programming and later moved into fintech domain and should be currently working as product manager in risk computation.
What I am getting is Data scientist/Business Analyst Profile which are not good as they don't have previous experience in coding. Budget is INR 2.7 to 3.3 Mn.
Interested? mail me for detailed JD on 1110mandar@gmail.com

**Comment by imrahul361** [comment]
Is it issue solvable for a beginner ? if yes can I work on it?

**Comment by Harsh188** [comment]
This issue should be closed, #41775 got merged and fixed the issue with the documentation.

**Comment by gittygupta** [comment]
in case someone is still stuck with NaN inputs, tf.where can be entirely replaced with:
```python
tf.minimum
```
